### PR TITLE
Check the chain for transactions if there are attempted burns that haven't been completed

### PR DESCRIPTION
**Pull request by maplant**

### CODE REVIEW

File: iot_packet_verifier/migrations/4_saved_balance.sql [hunk 1]
```diff
@@ -0,0 +1,4 @@
+CREATE TABLE saved_balances (
+        payer TEXT PRIMARY KEY,
+        balance BIGINT NOT NULL
```
Comment (andymck) [inline iot_packet_verifier/migrations/4_saved_balance.sql:3]: Be good to have a last_update or insert_date here

File: iot_packet_verifier/src/pending_burns.rs [hunk 2]
```diff
@@ -216,3 +332,22 @@ pub struct Burn {
     pub payer: PublicKeyBinary,
     pub amount: i64,
 }
+
+#[derive(Clone, FromRow, Debug)]
+pub struct BurnAttempt {
+    pub payer: PublicKeyBinary,
+    amount: i64,
+    latest_transaction_signature: String,
```
Comment (andymck) [inline iot_packet_verifier/src/pending_burns.rs:340]: insert date would be useful here, even if only to assist manual auditing, verification when browsing the table

File: solana/src/lib.rs [hunk 3]
```diff
@@ -227,6 +248,73 @@ impl SolanaNetwork for SolanaRpc {
 
         Ok(())
     }
+
+    async fn latest_transaction(&self) -> Result<Signature, Self::Error> {
+        let signer = Keypair::from_bytes(&self.keypair).unwrap();
```
Comment (andymck) [inline solana/src/lib.rs:253]: this unwrap safe?

File: iot_packet_verifier/migrations/4_saved_balance.sql [hunk 4]
```diff
@@ -0,0 +1,4 @@
+CREATE TABLE saved_balances (
```
Comment (andymck) [inline iot_packet_verifier/migrations/4_saved_balance.sql:1]: is this table redundant now?

File: iot_packet_verifier/src/burner.rs [hunk 5]
```diff
@@ -58,6 +61,42 @@ where
...
+            tracing::info!(%since, %amount, "Checking chain for burn transaction");
+            if self
+                .solana
+                .has_burn_transaction(&attempted_burn.payer, amount, &since)
+                .await
+                .map_err(BurnError::SolanaError)?
+            {
+                tracing::info!("Found a matching transaction. Removing burn attempt");
+                self.pending_burns
+                    .complete_burn_attempt(&attempted_burn.payer, amount)
```
Comment (andymck) [inline iot_packet_verifier/src/burner.rs:84]: On the regular burn function, the payer balance is updated after a complete_burn_attempt is successful. Does the same need performed here after this run of complete_burn_attempt?

File: solana/src/lib.rs (hunk 3 above)
Comment (maplant) [reply solana/src/lib.rs:253]: Yup, we do this in the burn function as well

File: iot_packet_verifier/migrations/4_saved_balance.sql (hunk 4 above)
Comment (maplant) [reply iot_packet_verifier/migrations/4_saved_balance.sql:1]: Yup, forgot to delete

File: Cargo.toml [hunk 6]
```diff
@@ -65,6 +65,7 @@ hextree = "*"
 solana-client = "1.14"
 solana-sdk = "1.14"
 solana-program = "1.11"
+solana-transaction-status = "1.14"
```
Comment (jeffgrunewald) [inline Cargo.toml:68]: friendly periodic reminder that the solana family of dependencies is in the 1.16.x range now. we'll have to struggle through an update eventually

File: iot_packet_verifier/src/pending_burns.rs [hunk 7]
```diff
@@ -48,25 +59,62 @@ impl PendingBurns for Pool<Postgres> {
             .await
     }
 
-    async fn subtract_burned_amount(
+    async fn fetch_incomplete_burns(&mut self) -> Result<Vec<BurnAttempt>, Self::Error> {
+        sqlx::query_as("SELECT * FROM attempted_burns")
```
Comment (andymck) [inline iot_packet_verifier/src/pending_burns.rs:63]: you forget to checkin the migration for this table ?

**Review by andymck — APPROVED** [review]
Other than the missing attempted_burns table migration, lgtm

**Pull request closed without merging**